tests/unit/model/relationships/rollback-test.js

```diff
@@ -0,0 +1,275 @@
+    });
+
+    env = setupStore({ person: Person, dog: Dog });
+    store = env.store;
+  }
+});
+
+test("saved changes to relationships should not roll back to a pre-saved state (from child)", function(assert) {
+  var person1, person2, dog1, dog2, dog3;
+
+  env.adapter.updateRecord = function(store, type, snapshot) {
+    return Ember.RSVP.resolve({ id: 2, owner: 1 });
+  };
+
+  run(function() {
+    store.push({
+      data: {
+        type: 'person',
+        id: 1,
+        attributes: {
+          firstName: "Tom",
+          lastName: "Dale"
+        },
+        relationships: {
+          dogs: {
+            data: [{
+              type: 'dog',
+              id: 1
+            }]
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'person',
+        id: 2,
+        attributes: {
+          firstName: "John",
+          lastName: "Doe"
+        },
+        relationships: {
+          dogs: {
+            data: [{
+              type: 'dog',
+              id: 2
+            }]
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 1,
+        attributes: {
+          name: "Fido"
+        },
+        relationships: {
+          owner: {
+            data: {
+              type: 'person',
+              id: 1
+            }
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 2,
+        attributes: {
+          name: "Bear"
+        },
+        relationships: {
+          owner: {
+            data: {
+              type: 'person',
+              id: 2
+            }
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 3,
+        attributes: {
+          name: "Spot"
+        },
+        relationships: {
+          owner: {
+            data: null
+          }
+        }
+      }
+    });
+    person1 = store.peekRecord('person', 1);
+    person2 = store.peekRecord('person', 2);
+    dog1 = store.peekRecord('dog', 1);
+    dog2 = store.peekRecord('dog', 2);
+    dog3 = store.peekRecord('dog', 3);
+
+    person1.get('dogs').addObject(dog2);
+  });
+
+  run(function() {
+    dog2.save().then(function () {
+      person1.get('dogs').addObject(dog3);
+      return Ember.RSVP.all([dog2.rollback(), dog3.rollback()]);
+    }).then(function () {
+      person1.get('dogs').then(function (dogs) {
+        assert.deepEqual(dogs.toArray(), [dog1,dog2]);
+      });
+      person2.get('dogs').then(function (dogs) {
+        assert.deepEqual(dogs.toArray(), []);
+      });
+      dog1.get('owner').then(function (owner) {
+        assert.equal(owner, person1);
+      });
+      dog2.get('owner').then(function (owner) {
+        assert.equal(owner, person1);
+      });
+    });
+  });
+});
+
+// skip("saved changes to relationships should not roll back to a pre-saved state (from parent)", function(assert) {
```

Comment:
The commented tests are for rolling back a hasMany relationship (via `parent.rollback()` which I don't think is proper behavior) instead of rolling back each modified child. Basically, I left it in here to get feedback from people on whether this behavior should be supported or not...

```js
// for something like:
//  parent: {
//    children: [ child1, child2 ]
//  }

parent.get('children').removeObject(child1)  // children: [ child2 ]
parent.get('children').addObject(child3)     // children: [ child2, child3 ]

parent.rollback() 
// some people believe this should result in `children: [ child1, child2 ]`;
// however the parent is not what is actually changing here
// so rolling it back doesn't really make sense; 
// child1 and child3 are actually what became dirty

[child1, child3].invoke('rollback'); // children: [ child1, child2 ]
```

This was sort of previously discussed in #2881...

Some have suggested to me that one should be able to rollback a hasMany relationship via the parent as described above. After relationships were refactored it became rather simple to rollback each childs' belongsTo but it is not obvious how to do the same from the parent's hasMany...

In order to rollback from the parent side it's not just as simple as reverting back to the relationship's canonical state... For example, there may have been unsaved objects added to that relationship. What would one do with those? Blindly delete them? What if they previously belonged to another parent? Do we blindly add them back to that other parents relationship? What if that other parent had been changed in the meantime and been saved, now what happens to that child?

I'm not really sure that supporting rollback from the parent in this case makes much sense. Each modified child basically has to be handled on a case by case basis anyway to achieve an expected result. The child is what points to - belongs to - a new parent, from a rollback perspective, a belongsTo relationship is very similar to an attributes and very easy to reason about. The same can't be said of the array of objects one has to work from the hasMany side of things...